nautobot_circuit_maintenance/jobs/site_search.py

```diff
@@ -0,0 +1,161 @@
+
+    data_dict = {
+        "msp": {
+            CM1,
+            CM2
+        },
+        "nyc": {
+            CM3,
+            CM4
+        }
+    }
+
+    Args:
+        maintenance_queryset (Queryset): Queryset of all Circuit Maintenance objects
+
+    Returns:
+        dict: Dictionary of a set of maintenance records
+    """
+    return_dictionary = collections.defaultdict(set)
+    for record in maintenance_queryset:
+        # Get the site overall
+        # Start with looping over the circuits that are defined
+        for circuit in record.circuits:
+            # Check both termination A and Z for values of None, add that record to the set for each site
+            for term in [circuit.termination_a, circuit.termination_z]:
+                if term is not None:
+                    return_dictionary[term.site.name].add(record)
+
+    return dict(return_dictionary)
+
+
+class FindSitesWithCircuitImpact(Job):
+    """Nautobot Job definition for finding sites without redundant circuit for impactful maintenance.
+
+    Args:
+        Job (Nautobot Job): Nautobot Job import.
+    """
+
+    class Meta:
+        """Meta definition for the Job."""
+
+        name = "Find Sites With Circuit Impact"
+        description = "Search for sites with multiple circuits, 1 or more circuit impacts."
+        read_only = True
+
+    def run(self, data=None, commit=None):
+        """Executes the Job.
+
+        Args:
+            data (_type_, optional): _description_. Defaults to None.
+            commit (_type_, optional): _description_. Defaults to None.
+        """
+        # Query for all of the circuits maintenances that are on going in the future
+        today = date.today()
+        circuit_maintenances = CircuitMaintenance.objects.filter(start_time__gte=today).order_by("start_time")
+
+        # Build a circuit mapper
+        circuit_maintenance_mapper = build_sites_to_maintenance_mapper(circuit_maintenances)
+
+        # Loop over each of the circuit maintenance records
+        for ckt_maint in circuit_maintenances:
+            # Get the list of sites
+            for circuit in ckt_maint.circuits:
+                sites = get_sites_from_circuit(circuit)
+
+            # Check to see if there are any circuit maintenances that are duplicated time
+            for site in sites:
+                for other_ckt_maint in circuit_maintenance_mapper[site.name]:
+                    # Report failures for any time where a circuit will take an outage
+                    if ckt_maint == other_ckt_maint:
+                        self.log_info("Found each other!")
+                        continue
+                    if check_for_overlap(ckt_maint, other_ckt_maint):
+                        self.log_warning(
+                            obj=site,
+                            message=f"There is an overlapping maintenance for site: {site.name}. Other maintenances: {other_ckt_maint}",
```

Comment:
should not document both maintenances? why only one?

Reply:
Trying to keep the length of the line short. Added.